src/pytorch/pme/pmeCPU.cpp

```diff
@@ -0,0 +1,383 @@
+            float mhx = mx*recipBoxVectors[0][0];
+            float bx = scaleFactor*xmod[kx];
+            for (int ky = 0; ky < gridSize[1]; ky++) {
+                int my = (ky < (gridSize[1]+1)/2) ? ky : ky-gridSize[1];
+                float mhy = mx*recipBoxVectors[1][0] + my*recipBoxVectors[1][1];
+                float mhx2y2 = mhx*mhx + mhy*mhy;
+                float bxby = bx*ymod[ky];
+                for (int kz = 0; kz < zsize; kz++) {
+                    int index = kx*yzsize + ky*zsize + kz;
+                    int mz = (kz < (gridSize[2]+1)/2) ? kz : kz-gridSize[2];
+                    float mhz = mx*recipBoxVectors[2][0] + my*recipBoxVectors[2][1] + mz*recipBoxVectors[2][2];
+                    float bz = zmod[kz];
+                    float m2 = mhx2y2 + mhz*mhz;
+                    float denom = m2*bxby*bz;
+                    float eterm = (index == 0 ? 0 : expf(-recipExpFactor*m2)/denom);
+                    float scale = (kz > 0 && kz <= (gridSize[2]-1)/2 ? 2 : 1);
+                    c10::complex<float>& g = recip[kx][ky][kz];
+                    energy += scale * eterm * (g.real()*g.real() + g.imag()*g.imag());
+                    g *= eterm;
+                }
+            }
+        }
+
+        // Store data for later use.
+
+        ctx->save_for_backward({positions, charges, box_vectors, xmoduli, ymoduli, zmoduli, recipGrid});
+        ctx->saved_data["gridx"] = gridx;
+        ctx->saved_data["gridy"] = gridy;
+        ctx->saved_data["gridz"] = gridz;
+        ctx->saved_data["order"] = order;
+        ctx->saved_data["alpha"] = alpha;
+        ctx->saved_data["coulomb"] = coulomb;
+        return {torch::tensor(0.5*energy, options)};
+    }
+
+    static tensor_list backward(AutogradContext *ctx, tensor_list grad_outputs) {
+        auto saved = ctx->get_saved_variables();
+        Tensor positions = saved[0];
+        Tensor charges = saved[1];
+        Tensor box_vectors = saved[2];
+        Tensor xmoduli = saved[3];
+        Tensor ymoduli = saved[4];
+        Tensor zmoduli = saved[5];
+        Tensor recipGrid = saved[6];
+        int gridSize[3] = {(int) ctx->saved_data["gridx"].toInt(), (int) ctx->saved_data["gridy"].toInt(), (int) ctx->saved_data["gridz"].toInt()};
+        int pmeOrder = (int) ctx->saved_data["order"].toInt();
+        float alpha = (float) ctx->saved_data["alpha"].toDouble();
+        float sqrtCoulomb = sqrt(ctx->saved_data["coulomb"].toDouble());
+        int numAtoms = positions.size(0);
+        auto box = box_vectors.accessor<float,2>();
+        auto pos = positions.accessor<float,2>();
+        auto charge = charges.accessor<float,1>();
+        float recipBoxVectors[3][3];
+        invertBoxVectors(box_vectors, recipBoxVectors);
+
+        // Take the inverse Fourier transform.
+
+        Tensor realGrid = torch::fft::irfftn(recipGrid, {gridSize[0], gridSize[1], gridSize[2]}, c10::nullopt, "forward");
```

Comment:
torch<2 refuses to compile this line because it cannot convert from ```{gridSize[0], gridSize[1], gridSize[2]}``` to the expected ```c10::optional<c10::ArrayRef<long int> >``` type. 
Maybe passing ``` int gridSizeTemp[3] = {gridSize[0], gridSize[1], gridSize[2]}``` makes it happy?